src/persist-client/src/cli/inspect.rs

```diff
@@ -439,6 +447,78 @@ pub async fn blob_counts(blob_uri: &str) -> Result<impl serde::Serialize, anyhow
     Ok(blob_counts)
 }
 
+/// Rummages through S3 to find the latest rollup for all shards, then calculates summary stats
+/// for each shard.
+pub async fn shard_stats(blob_uri: &str) -> anyhow::Result<()> {
+    let cfg = PersistConfig::new(&READ_ALL_BUILD_INFO, SYSTEM_TIME.clone());
+    let metrics = Arc::new(Metrics::new(&cfg, &MetricsRegistry::new()));
+    let blob = make_blob(&cfg, blob_uri, NO_COMMIT, metrics).await?;
+
+    // Collect the latest rollup for every shard with the given blob_uri
+    let mut rollup_keys = BTreeMap::new();
+    blob.list_keys_and_metadata(&BlobKeyPrefix::All.to_string(), &mut |metadata| {
+        if let Ok((shard, PartialBlobKey::Rollup(seqno, rollup_id))) =
+            BlobKey::parse_ids(metadata.key)
+        {
+            let key = (seqno, rollup_id);
+            match rollup_keys.entry(shard) {
+                Entry::Vacant(v) => {
+                    v.insert(key);
+                }
+                Entry::Occupied(o) => {
+                    if key.0 > o.get().0 {
+                        *o.into_mut() = key;
+                    }
+                }
+            };
+        }
+    })
+    .await?;
+
+    println!("parts,runs,batches,longest_run,byte_width");
+    for (shard, (seqno, rollup)) in rollup_keys {
+        let rollup_key = PartialRollupKey::new(seqno, &rollup).complete(&shard);
+        // Basic stats about the trace.
+        let mut parts = 0;
+        let mut runs = 0;
+        let mut batches = 0;
+        let mut longest_run = 0;
+        // The sum of the largest part in every run, measured in bytes.
+        // A rough proxy for the worst-case amount of data we'd need to fetch to consolidate
+        // down a single key-value pair.
+        let mut byte_width = 0;
+
+        let Some(rollup) = blob.get(&rollup_key).await? else {
+            // Deleted between listing and now?
+            continue;
+        };
+
+        let state: State<u64> =
+            UntypedState::decode(&cfg.build_version, &rollup).check_ts_codec(&shard)?;
+        state.collections.trace.map_batches(|b| {
+            parts += b.parts.len();
+            runs += if b.parts.is_empty() {
```

Comment:
preference for `b.runs().count()` to keep the logic in one place. it does a bit more work, but seems fine given the context

also, I haven't thought about this too hard, but seems like it's what you want below, too? in which case, it's not even more work

Reply:
Yes, that does look like exactly what I want! I figured something like this might exist but didn't find it. (Okay if i move out of `compact.rs` and next to the struct implementation?)

Reply:
yup!